features/AddKudo.go

```diff
@@ -16,30 +16,31 @@ type AddKudo struct {
 	kudoCounts model.KudoCounts
 }
 
+var lastUpdate = make(map[int64]tgbotapi.Update)
+
 func NewAddKudo(kudosService model.Kudos, userService model.Users, kudoCountService model.KudoCounts) AddKudo {
+
 	return AddKudo{kudos: kudosService, users: userService, kudoCounts: kudoCountService}
 }
 
 func (k AddKudo) Execute(update tgbotapi.Update, db *gorm.DB, bot *tgbotapi.BotAPI) {
+	if lastUpdate[update.Message.Chat.ID].Message == nil {
+		return
+	}
 	var msg tgbotapi.MessageConfig
-	receiver, err := k.users.CreateUserIfNotExist(update.Message.ReplyToMessage.From, db)
+	_, err := k.users.CreateUserIfNotExist(update.Message.From, db)
 	if err != nil && err != gorm.ErrRecordNotFound {
 		msg = tgbotapi.NewMessage(update.Message.Chat.ID, fmt.Sprintf("Kudo error: %s", err))
 	}
 
-	_, err = k.users.CreateUserIfNotExist(update.Message.From, db)
+	receiver, err := k.users.CreateUserIfNotExist(lastUpdate[update.Message.Chat.ID].Message.From, db)
+	if err != nil && err != gorm.ErrRecordNotFound {
 
-	// You may not vote on your own message.
-	if strings.EqualFold(update.Message.ReplyToMessage.Text, "+") || strings.EqualFold(update.Message.ReplyToMessage.Text, "-") {
-		err = errors.New("voting on a kudo not allowed")
 		msg = tgbotapi.NewMessage(update.Message.Chat.ID, fmt.Sprintf("Kudo error: %s", err))
-		_, err = bot.Send(msg)
-		if err != nil {
-			log.Printf("error sending message %s\n", err)
-		}
-		return
+
 	}
-	if update.Message.From.ID == update.Message.ReplyToMessage.From.ID {
+
+	if lastUpdate[update.Message.Chat.ID].Message.From.ID == update.Message.From.ID {
```

Comment:
Ik zou dit haast een memory leak noemen want je slaat nu eigenlijk alle messages in memory op toch?

Waarom werkte de vorige oplossing niet? `update.Message.From.ID == update.Message.ReplyToMessage.From.ID` dat leek vrij logisch.

Reply:
Ooooooh, ik zie het. Hiermee kan je + roepen zonder reply en dan is het een vote op het laatste bericht?

Waarom bewaar je dan alle berichten? Je zou toch gewoon het laatste bericht kunnen bewaren dat geen kudo was? (granted, de architectuur leent zich er niet voor).

Reply:
Je zou eventueel in je bovenste laag een "previousUntriggeredMessage" bij kunnen houden die je via een context aan je Action doorgeeft.